File: cal-reducer-go/migrate.go
package calreducer

// PFC2-M3 — multisig (AuthorizationSet v2) registry: the v1→v2 migration and the well-formed
// owner-record invariant. Mirrors migrate.ts / migrate.rs (§1.1/§4). The bound is enforced where
// Deltas commit an owner record (see delta.go); this file is the single source of the invariant and
// the deterministic upgrade (state-only, no external input ⇒ reproducible for the M5 vectors).

import (
	"sort"
	"strconv"

	canonical "github.com/paradigm-terra/canonical-go"
)

// MaxOwners is the §1.1 upper bound on the owner set.
const MaxOwners = 16

// OwnerRecordWellFormed is the §1.1 invariant: owners non-empty, DISTINCT, ascending (by raw pubkey
// bytes; equal-length hex, so string order == byte order), at most MaxOwners; and
// 1 <= threshold <= len(owners). Mirrors ownerRecordWellFormed.
func OwnerRecordWellFormed(owners, threshold canonical.Value) bool {
	arr, ok := owners.([]canonical.Value)
	if !ok {
		return false
	}
	tInt, ok := threshold.(canonical.Int)
	if !ok {
		return false
	}
	t, err := strconv.ParseUint(string(tInt), 10, 64)
	if err != nil {
		return false
	}
	n := len(arr)
	if n < 1 || n > MaxOwners {
		return false
	}
	if t < 1 || t > uint64(n) {
		return false
	}
	for i := range n {
		o, ok := arr[i].(string)
		if !ok || o == "" {
			return false
		}
		if i > 0 {
			prev, _ := arr[i-1].(string)
			if o == prev || o < prev { // distinct + ascending
				return false
			}
		}
	}
	return true
}

// MigrateRegistryV1ToV2 is the §4 deterministic v1→v2 registry upgrade — a pure function of the
// state alone (no external input), idempotent. owner_pubkey:"K" → owners:["K"], threshold:1 (1-of-1
// bridge, SC-4); owner_pubkey:"" → owners:[], threshold:0 (no-owner). Mirrors migrateRegistryV1ToV2.
func MigrateRegistryV1ToV2(state canonical.Value) canonical.Value {
	agentsV, ok := getIn(state, []string{"registry", "agents"})
	if !ok {
		return state
	}
	agents, ok := agentsV.(*canonical.Object)
	if !ok {
		return state
	}
	ids := append([]string(nil), agents.Keys()...)
	sort.Strings(ids)
	s := state
	for _, id := range ids {
		recV, ok := getIn(s, []string{"registry", "agents", id})
		if !ok {
			continue
		}
		rec, ok := recV.(*canonical.Object)
		if !ok {
			continue
		}
		if _, hasOwners := rec.Get("owners"); hasOwners {
			continue // already v2
		}
		pkV, hasPk := rec.Get("owner_pubkey")
		if !hasPk {
			continue // nothing to migrate
		}
		pk, _ := pkV.(string)
		owners := []canonical.Value{}
		threshold := "0"
		if pk != "" {
			owners = []canonical.Value{pk}
			threshold = "1"
		}
		s = setIn(s, []string{"registry", "agents", id, "owners"}, owners)
		s = setIn(s, []string{"registry", "agents", id, "threshold"}, canonical.Int(threshold))
		s = deleteIn(s, []string{"registry", "agents", id, "owner_pubkey"})
	}
	return s
}

// enforceOwnerRecord is the PFC2-M3 §1.1 bound, enforced where a Delta commits
// registry/agents/<a>/owners or .../threshold. Fires only once BOTH fields are present (a
// partially-built record is a transient intermediate). Mirrors enforceOwnerRecord in delta.ts.
func enforceOwnerRecord(state canonical.Value, full []string) *ApplyError {
	if len(full) != 4 || full[0] != "registry" || full[1] != "agents" {
		return nil
	}
	if full[3] != "owners" && full[3] != "threshold" {
		return nil
	}
	recV, ok := getIn(state, []string{"registry", "agents", full[2]})
	if !ok {
		return nil
	}
	rec, ok := recV.(*canonical.Object)
	if !ok {
		return nil
	}
	ownersV, hasOwners := rec.Get("owners")
	thV, hasTh := rec.Get("threshold")
	if !hasOwners || !hasTh {
		return nil // not yet complete
	}
	if !OwnerRecordWellFormed(ownersV, thV) {
		return aerr("BAD_OWNER_RECORD")
	}
	return nil
}

File: cal-reducer-rs/src/migrate.rs
//! PFC2-M3 — multisig (AuthorizationSet v2) registry: the v1→v2 migration and the well-formed
//! owner-record invariant. Mirrors `migrate.ts` (§1.1/§4). The bound is enforced where Deltas
//! commit an owner record (see `delta.rs`); this module is the single source of the invariant and
//! the deterministic upgrade (state-only, no external input ⇒ reproducible for the M5 vectors).

use paradigm_terra_canonical::jcs::JcsValue;

use crate::state::{delete_in, get_in, set_in};

/// §1.1 upper bound on the owner set.
pub const MAX_OWNERS: usize = 16;

/// §1.1: a well-formed v2 owner record — `owners` non-empty, DISTINCT, ascending (by raw pubkey
/// bytes; equal-length hex strings, so string order == byte order), at most MAX_OWNERS; and
/// `1 <= threshold <= owners.len()`. Mirrors `ownerRecordWellFormed`.
pub fn owner_record_well_formed(owners: Option<&JcsValue>, threshold: Option<&JcsValue>) -> bool {
    let owners = match owners.and_then(JcsValue::as_array) {
        Some(a) => a,
        None => return false,
    };
    let t: u64 = match threshold {
        Some(JcsValue::Int(s)) => match s.parse() {
            Ok(n) => n,
            Err(_) => return false,
        },
        _ => return false,
    };
    let n = owners.len();
    if n < 1 || n > MAX_OWNERS {
        return false;
    }
    if t < 1 || t > n as u64 {
        return false;
    }
    for i in 0..n {
        let o = match owners[i].as_str() {
            Some(s) if !s.is_empty() => s,
            _ => return false,
        };
        if i > 0 {
            let prev = owners[i - 1].as_str().unwrap_or("");
            if o == prev || o < prev {
                return false; // distinct + ascending
            }
        }
    }
    true
}

/// §4: the deterministic v1→v2 registry upgrade — a pure function of the state alone (no external
/// input), idempotent. `owner_pubkey:"K"` → `owners:["K"], threshold:1` (1-of-1 bridge, SC-4);
/// `owner_pubkey:""` → `owners:[], threshold:0` (no-owner record). Mirrors `migrateRegistryV1ToV2`.
pub fn migrate_registry_v1_to_v2(state: &JcsValue) -> JcsValue {
    let ids: Vec<String> = match get_in(state, &["registry", "agents"]) {
        Some(JcsValue::Object(a)) => {
            let mut v: Vec<String> = a.iter().map(|(k, _)| k.clone()).collect();
            v.sort();
            v
        }
        _ => return state.clone(),
    };
    let mut s = state.clone();
    for id in ids {
        let (is_v2, has_owner_pubkey, pk) = {
            let rec = match get_in(&s, &["registry", "agents", &id]) {
                Some(r) => r,
                None => continue,
            };
            let is_v2 = rec.get("owners").and_then(JcsValue::as_array).is_some();
            let has_op = rec.get("owner_pubkey").is_some();
            let pk = rec.get("owner_pubkey").and_then(JcsValue::as_str).unwrap_or("").to_string();
            (is_v2, has_op, pk)
        };
        if is_v2 || !has_owner_pubkey {
            continue;
        }
        let (owners, threshold) = if pk.is_empty() {
            (Vec::new(), "0")
        } else {
            (vec![JcsValue::Str(pk)], "1")
        };
        s = set_in(&s, &["registry", "agents", &id, "owners"], JcsValue::Array(owners));
        s = set_in(&s, &["registry", "agents", &id, "threshold"], JcsValue::Int(threshold.to_string()));
        s = delete_in(&s, &["registry", "agents", &id, "owner_pubkey"]);
    }
    s
}
